---
title: "scRNA-seq analysis (Filtered)"
author: "Pedro R"
date: "14 January 2020"
output:
  html_document: default
  df_print: paged
params:
  input_path: input_path
  is_test: is_test
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Read arguments and loadings

```{r loading}
options(warn=-1)

aggregated_path = params$input_path

library(Seurat)
library(dplyr)
library(mvoutlier)

load(file = aggregated_path)
```

## Normalize

```{r normalize}
aggregated_samples = PercentageFeatureSet(aggregated_samples, pattern = "^MT-", col.name = "percent.mt")
aggregated_samples = SCTransform(aggregated_samples, verbose = FALSE)
aggregated_samples = RunPCA(aggregated_samples)
```

## Find principal components

```{r components}
## jackstraw does not work due to breaking change in 3.1.5
## see https://github.com/sansomlab/tenx/issues/100
 aggregated_samples = JackStraw(aggregated_samples, num.replicate = 100, dims = 50, verbose = FALSE)
 aggregated_samples = ScoreJackStraw(aggregated_samples, dims = 1:50)
 jackstraw_pvalues = data.frame(aggregated_samples@reductions$pca@jackstraw$overall.p.values)

for (i in 1:dim(jackstraw_pvalues)[1]) {
  if (jackstraw_pvalues$Score[i] > 0.05) {
    PC_number = i - 1
    break
  }
}

ElbowPlot(aggregated_samples, ndims = 50)
JackStrawPlot(aggregated_samples, dims = 1:50)
```

## Clustering and remove dead cells

```{r clustering_and_remove_dead_cells}
aggregated_samples = FindNeighbors(aggregated_samples, dims = 1:PC_number)
aggregated_samples = FindClusters(aggregated_samples)
VlnPlot(aggregated_samples, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), group.by = 'orig.ident', ncol = 3, pt.size = 0)

cells <- tryCatch({
  cells = c()
  x = aggregated_samples@meta.data[ , c("nFeature_RNA","nCount_RNA","percent.mt")]
  outliers = aq.plot(x, delta=qchisq(0.975, df = ncol(x)))
  outliers_index = which(unlist(outliers))
  outliers_names = rownames(x)[outliers_index]
  all_cells = WhichCells(aggregated_samples)
  cells = setdiff(all_cells, outliers_names)
  return(cells)
  },
  error = function(err) {
    cells = c()
    return(cells)
    })
dev.off()
gc()

```

## Reload data

```{r reload}
load(file = aggregated_path)
aggregated_samples = SubsetData(aggregated_samples, cells = cells)
```

## Normalize with cell cycle effect removal

```{r normalize2}
aggregated_samples = PercentageFeatureSet(aggregated_samples, pattern = "^MT-", col.name = "percent.mt")

s.genes <- cc.genes.updated.2019$s.genes
g2m.genes <- cc.genes.updated.2019$g2m.genes

### only run cell cycle scoring with real data
if (params$is_test) {
  aggregated_samples = SCTransform(aggregated_samples, verbose = FALSE)
} else {
  aggregated_samples <- CellCycleScoring(aggregated_samples, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
  aggregated_samples$CC.Difference <- aggregated_samples$S.Score - aggregated_samples$G2M.Score
  aggregated_samples = SCTransform(aggregated_samples, vars.to.regress = "CC.Difference", verbose = FALSE)
}

aggregated_samples = RunPCA(aggregated_samples)
```

## Find principal components

```{r components2}
aggregated_samples = JackStraw(aggregated_samples, num.replicate = 100, dims = 50, verbose = FALSE)
aggregated_samples = ScoreJackStraw(aggregated_samples, dims = 1:50)
jackstraw_pvalues = data.frame(aggregated_samples@reductions$pca@jackstraw$overall.p.values)

for (i in 1:dim(jackstraw_pvalues)[1]) {
  if (jackstraw_pvalues$Score[i] > 0.05) {
    PC_number = i - 1
    break
  }
}

ElbowPlot(aggregated_samples, ndims = 50)
JackStrawPlot(aggregated_samples, dims = 1:50)
```

## Clustering and visualize

```{r clustering2}
aggregated_samples = FindNeighbors(aggregated_samples, dims = 1:PC_number)
aggregated_samples = FindClusters(aggregated_samples)
VlnPlot(aggregated_samples, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), group.by = 'orig.ident', ncol = 3, pt.size = 0)

aggregated_samples <- RunTSNE(aggregated_samples, dims = 1:PC_number, check_duplicates = FALSE)
aggregated_samples <- RunUMAP(aggregated_samples, dims = 1:PC_number)

DimPlot(aggregated_samples, reduction = "pca", group.by = "orig.ident")
DimPlot(aggregated_samples, reduction = "tsne", group.by = "orig.ident")
DimPlot(aggregated_samples, reduction = "umap", group.by = "orig.ident")
FeaturePlot(aggregated_samples, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"))
```

## Differential expression

```{r differential_expression}
dummy <- tryCatch({
    Idents(object = aggregated_samples) <- "orig.ident"
    list_sample_names = as.vector(unique((aggregated_samples@active.ident)))
    for (sample in list_sample_names) {
      markers <- FindMarkers(aggregated_samples, ident.1 = sample)
      print(sample)
      head(markers, n = 10)
    }
    aggregated_samples_markers = FindAllMarkers(aggregated_samples, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
    top10 = aggregated_samples_markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
    DoHeatmap(aggregated_samples, features = top10$gene) + NoLegend()
    result = "NA"
    return(result)
  },
  error = function(err){
    writeLines("WARNING: **No Features found**")
    result = "NA"
    return(result)
    })
```

## Save file

```{r saving}
save(aggregated_samples, file = "aggregated_object_analyzed_filtered.RData")
```
